base/linalg/generic.jl

```diff
@@ -103,21 +103,33 @@ function generic_vecnorm1(x)
     return convert(T, sum)
 end
 
+function generic_vecnorm2(n, scale, ssq, x)
+    @inbounds for i = 1:n
+        if x[i] != 0
+            absxi = norm(x[i])
+            if scale < absxi
+                ssq = 1 + ssq*(scale/absxi)^2
+                scale = absxi
+            else
+                ssq = ssq + (absxi/scale)^2
+            end
+        end
+    end
+    return scale * sqrt(ssq)
+end
+function generic_vecnorm2{T<:Number}(x::AbstractVector{T})
+    scale = zero(norm(zero(T)))
+    ssq = one(sqrt((abs2(norm(zero(T))) + abs2(norm(zero(T))))/2))
```

Comment:
It almost is for `Float64`:

```jl
julia> f{T}(x::T) = one(sqrt((abs2(norm(zero(T))) + abs2(norm(zero(T))))/2))
f (generic function with 2 methods)

julia> f(123.45)
1.0

julia> @code_llvm f(123.45)

define double @julia_f_21321(double) {
top:
  %1 = call double @julia_norm_21171(double 0.000000e+00)
  %2 = call double @julia_norm_21171(double 0.000000e+00)
  ret double 1.000000e+00
}
```

So it figures out that the result is constant but still calls the Float64 norm function twice in case there are side effects.